File: pytorch-plugin/src/main/java/fr/greenpydata/pytorch/checks/AvoidNonPinnedMemoryForDataloaders.java
package fr.greenpydata.pytorch.checks;

import org.sonar.check.Priority;
import org.sonar.check.Rule;
import org.sonar.plugins.python.api.PythonSubscriptionCheck;
import org.sonar.plugins.python.api.tree.Tree;
import org.sonar.plugins.python.api.tree.CallExpression;
import org.sonar.plugins.python.api.tree.RegularArgument;
import org.sonar.plugins.python.api.tree.Expression;
import org.sonar.plugins.python.api.tree.Name;

import java.util.Optional;

import static org.sonar.plugins.python.api.tree.Tree.Kind.NAME;

@Rule(
  key = AvoidNonPinnedMemoryForDataloaders.RULE_KEY,
  name = AvoidNonPinnedMemoryForDataloaders.MESSAGE,
  description = AvoidNonPinnedMemoryForDataloaders.MESSAGE,
  priority = Priority.MINOR,
  tags = {"bug", "eco-design"})
public class AvoidNonPinnedMemoryForDataloaders extends PythonSubscriptionCheck {

  public static final String RULE_KEY = "P3";
  private static final String dataloaderFullyQualifiedName = "torch.utils.data.DataLoader";
  private static final int pinMemoryArgumentPosition = 7;
  private static final String pinMemoryArgumentName = "pin_memory";
  protected static final String MESSAGE = "Use pinned memory to reduce data transfer in RAM.";

  @Override
  public void initialize(Context context) {
    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, ctx -> {
      CallExpression callExpression = (CallExpression) ctx.syntaxNode();

      if (Utils.getQualifiedName(callExpression).equals(dataloaderFullyQualifiedName)) {
        RegularArgument numWorkersArgument = Utils.nthArgumentOrKeyword(pinMemoryArgumentPosition,
          pinMemoryArgumentName,
          callExpression.arguments());

        if (numWorkersArgument == null) {
          ctx.addIssue(callExpression, MESSAGE);
        } else {
          Optional.of(numWorkersArgument).filter(this::checkBadValuesForPinMemory)
            .ifPresent(arg -> ctx.addIssue(arg, MESSAGE));
        }
      }
    });
  }

  private boolean checkBadValuesForPinMemory(RegularArgument pinMemoryArgument) {
    Expression expression = pinMemoryArgument.expression();
    return expression.is(NAME) && ((Name) expression).name().equals("False");
  }
}

File: pytorch-plugin/src/test/resources/checks/AvoidNonPinnedMemoryForDataloaders.py
import torch
from torch.utils.data import DataLoader
from torch.utils.data import DataLoader as DL
import torch.utils as utils
import nottorch

dl = torch.utils.data.DataLoader(dataset) # Noncompliant {{Use pinned memory to reduce data transfer in RAM.}}
dl = torch.utils.data.DataLoader(dataset, num_workers=3, batch_size=1, shuffle=False, pin_memory=False) # Noncompliant {{Use pinned memory to reduce data transfer in RAM.}}
dl = DataLoader(dataset, batch_size=1, num_workers=0, shuffle=True, pin_memory=False) # Noncompliant {{Use pinned memory to reduce data transfer in RAM.}}
dl = torch.utils.data.DataLoader(num_workers=5, batch_size=2, shuffle=True) # Noncompliant {{Use pinned memory to reduce data transfer in RAM.}}
dl = DataLoader(dataset, 1, False, None, None, 0, None, False, False) # Noncompliant {{Use pinned memory to reduce data transfer in RAM.}}
dl = utils.data.DataLoader(dataset, batch_size=1, False, None, None, 0, None, False, True) # Noncompliant {{Use pinned memory to reduce data transfer in RAM.}}
dl = utils.data.DataLoader(dataset, pin_memory=False) # Noncompliant {{Use pinned memory to reduce data transfer in RAM.}}
dl = DL(dataset, pin_memory=False) # Noncompliant {{Use pinned memory to reduce data transfer in RAM.}}

dl = torch.utils.data.DataLoader(dataset, num_workers=3, batch_size=1, shuffle=False, pin_memory=True)
dl = DataLoader(dataset, num_workers=0, batch_size=1, shuffle=False, pin_memory=True)
dl = DataLoader(dataset, num_workers=0, batch_size=1, shuffle=False, pin_memory=True)
dl = utils.data.DataLoader(dataset, pin_memory=True)
dl = utils.data.DataLoader(dataset, batch_size=1, False, None, None, 0, None, True, True)
dl = DataLoader(dataset, batch_size=1, False, None, None, 0, None, True, False)
dl = torch.utils.data.DataLoader(torchvision.datasets.MNIST('/files/', train=True, download=True), pin_memory=True)
dl = DL(dataset, pin_memory=True)

dl = nottorch.utils.data.DataLoader(dataset, pin_memory=True)
dl = nottorch.utils.data.DataLoader(dataset, pin_memory=False)